src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs

```diff
@@ -13,95 +13,46 @@ namespace Microsoft.Extensions.Caching.Memory
+        private IList<IChangeToken> _expirationTokens;
         private TimeSpan? _slidingExpiration;
         private long? _size;
         private IDisposable _scope;
         private object _value;
-        private bool _valueHasBeenSet;
+        private int _state; // actually a [Flag] enum called "State"
 
-        internal readonly object _lock = new object();
-
-        internal CacheEntry(
-            object key,
-            Action<CacheEntry> notifyCacheEntryCommit,
-            Action<CacheEntry> notifyCacheOfExpiration,
-            ILogger logger)
+        internal CacheEntry(object key, MemoryCache memoryCache)
         {
-            if (key == null)
-            {
-                throw new ArgumentNullException(nameof(key));
-            }
-
-            if (notifyCacheEntryCommit == null)
-            {
-                throw new ArgumentNullException(nameof(notifyCacheEntryCommit));
-            }
-
-            if (notifyCacheOfExpiration == null)
-            {
-                throw new ArgumentNullException(nameof(notifyCacheOfExpiration));
-            }
-
-            if (logger == null)
-            {
-                throw new ArgumentNullException(nameof(logger));
-            }
-
-            Key = key;
-            _notifyCacheEntryCommit = notifyCacheEntryCommit;
-            _notifyCacheOfExpiration = notifyCacheOfExpiration;
-
+            Key = key ?? throw new ArgumentNullException(nameof(key));
+            _cache = memoryCache ?? throw new ArgumentNullException(nameof(memoryCache));
             _scope = CacheEntryHelper.EnterScope(this);
-            _logger = logger;
         }
 
         /// <summary>
         /// Gets or sets an absolute expiration date for the cache entry.
         /// </summary>
-        public DateTimeOffset? AbsoluteExpiration
-        {
-            get
-            {
-                return _absoluteExpiration;
-            }
-            set
-            {
-                _absoluteExpiration = value;
-            }
-        }
+        public DateTimeOffset? AbsoluteExpiration { get; set; }
 
         /// <summary>
         /// Gets or sets an absolute expiration time, relative to now.
         /// </summary>
         public TimeSpan? AbsoluteExpirationRelativeToNow
         {
-            get
-            {
-                return _absoluteExpirationRelativeToNow;
-            }
+            get => AbsoluteExpiration.HasValue ? AbsoluteExpiration.Value - _cache.GetUtcNow() : null;
             set
             {
-                if (value <= TimeSpan.Zero)
+                if (value.HasValue)
                 {
-                    throw new ArgumentOutOfRangeException(
-                        nameof(AbsoluteExpirationRelativeToNow),
-                        value,
-                        "The relative expiration value must be positive.");
+                    AbsoluteExpiration = value.Value > TimeSpan.Zero
+                        ? (_cache.GetUtcNow() + value)
```

Comment:
This is a subtle behavior change. Previously, when someone used `AbsoluteExpirationRelativeToNow`, the time didn't start until `MemoryCache.SetEntry` was called. (i.e. when the CacheEntry was disposed).

However, here, the time starts when the caller set the `AbsoluteExpirationRelativeToNow` property.

The difference can be shown with the following:

```C#
if (!cache.TryGetValue(key, out cachedValue))
{
    using (var entry = cache.CreateEntry(key))
    {
        entry.AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(1);  // the clock "starts" here now

        // some long running operation to materialize the value, maybe make a web request

        entry.Value = newValue;     
    }  // The clock didn't "start" until here previously
}

Reply:
@eerhardt you are right. However, I am not sure what the user would expect: the time relative to the moment when the property was set or to time when it was actually added to the cache?

I have reverted this change as it was also... causing a performance regression because clock was now called twice: first time when `AbsoluteExpirationRelativeToNow` was set and 2nd when `SetEntry` was calling it. I've added a benchmark to perf repo to ensure that such regression would not go unnoticed: https://github.com/dotnet/performance/pull/1606